src/sdk/service/network-validation-service/network-validation-service.ts

```diff
@@ -0,0 +1,468 @@
+   */
+  public static getSubnetCidr(netmask: string, ipAddress: string): string | null {
+    const prefixSize = this.getNetmaskPrefixSize(netmask);
+    if (!this.isValidIPv4Address(ipAddress) || prefixSize === -1) {
+      return null;
+    }
+    const components = ipAddress.split('.');
+    const prefixMod = Math.floor((prefixSize - 1) / 8);
+    let subnetIp = '';
+    for (let i = 0; i < 4; i++) {
+      const val = components[i];
+      if (i < Math.ceil(prefixMod)) {
+        subnetIp += components[i];
+      } else {
+        subnetIp += parseInt(val, 10) & (-1 << (8 * (i + 1)) - prefixSize);
+      }
+      if (i !== 3) {
+        subnetIp += '.';
+      }
+    }
+    return subnetIp + '/' + prefixSize;
+  }
+
+  /**
+   * Gets the range of a subnet from the netmask and any IP address on the
+   * network. If either the netmask or IP address are invalid, returns null.
+   * @param {string} netmask: the netmask string.
+   * @param {string} ipAddress: the IP address string.
+   * @return {IPv4AddressRange} an object of the form { start: "x.x.x.x", end: "x.x.x.x" },
+   *         where the start field contains the IP address at the low end of the
+   *         range and the end field contains the IP address at the high end of
+   *         the range.
+   */
+  public static getSubnetIPv4AddressRange(netmask: string, ipAddress: string): IPv4AddressRange | null {
+    if (!this.isValidIPv4Address(ipAddress) || !this.isValidNetmask(netmask)) {
+      return null;
+    }
+    const combined = this.getIPv4AsInteger(ipAddress);
+    const combinedMask = this.getIPv4AsInteger(netmask);
+    const low = combined & combinedMask;
+    const high = low | (~combinedMask);
+    return {
+      start: this.getIntegerAsIPv4Address(low),
+      end: this.getIntegerAsIPv4Address(high)
+    };
+  }
+
+  /**
+   * Gets a string representation of an IP address range. Returns null if either
+   * the netmask or ipAddress are invalid.
+   * @param {string} netmask: the netmask string.
+   * @param {string} ipAddress: the IP address string.
+   * @return {string} a string representation of the IP address range. For
+   *         example, "192.168.0.1 - 192.168.0.10".
+   */
+  public static getSubnetIPv4AddressRangeStr(netmask: string, ipAddress: string): string | null {
+    const range = this.getSubnetIPv4AddressRange(netmask, ipAddress);
+    if (!range) {
+      return null;
+    }
+    return range.start + '-' + range.end;
+  }
+
+  /**
+   * Gets a string representation of an IP address range without the .0 and .255 if applicable. Returns null if either
+   * the netmask or ipAddress are invalid.
+   * @param {string} netmask: the netmask string.
+   * @param {string} ipAddress: the IP address string.
+   * @return {string} a string representation of the IP address range. For
+   *         example, "192.168.0.1 - 192.168.0.10".
+   */
+  public static getSubnetIPv4AddressRangeStrWithoutReserved(netmask: string, ipAddress: string): string | null {
```

Comment:
it would be better to combine this and the preceding method and simply include a boolean value as the last param that indicates whether reserved values should be included